iwyu_ast_util.cc

```diff
@@ -546,32 +560,32 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
 
   // --- Methods on RecursiveASTVisitor
   bool TraverseType(QualType type) {
-    TraverseArgumentsOfSugaredTemplates(type);
+    TraverseTypeHelper(type);
     return Base::TraverseType(type);
   }
 
   bool TraverseTypeLoc(TypeLoc type_loc) {
-    TraverseArgumentsOfSugaredTemplates(type_loc.getType());
+    TraverseTypeHelper(type_loc.getType());
     return Base::TraverseTypeLoc(type_loc);
   }
 
-  bool VisitType(Type* type) {
-    seen_types_.insert(type);
-    return true;
-  }
-
  private:
-  // Clang doesn't traverse underlying type for most of the sugar types.
-  // If a TemplateSpecializationType occurs in a sugaring chain, traverse its
-  // arguments explicitly, otherwise they aren't put into resugar_map.
-  void TraverseArgumentsOfSugaredTemplates(QualType type) {
-    if (type.isNull())
+  void TraverseTypeHelper(QualType qual_type) {
+    if (qual_type.isNull())
       return;
 
-    if (const auto* template_spec = type->getAs<TemplateSpecializationType>()) {
-      for (const TemplateArgument& arg : template_spec->template_arguments())
-        TraverseTemplateArgument(arg);
-    }
+    const Type* const type = qual_type.getTypePtr();
+    const Type* const desugared_until_alias_or_tpl = Desugar(type);
+    seen_types_.insert(desugared_until_alias_or_tpl);
+
+    const Type* const desugared =
+        DesugarAliasTypes(desugared_until_alias_or_tpl);
+    // A desugared type shouldn't be inserted into resugar_map along with its
+    // sugared version as a value instead of a key, hence it is important to
```

Comment:
What does `along with its sugared version as a value instead of a key` mean? I suspect there's a small typo in this comment that makes it hard to understand...?

Reply:
`resugar_map` maps desugared types to their original sugared versions, obviously. So, desugared (canonical) types are the keys, and (probably) sugared types are the values. `TypeEnumerator` prepares _values_ for the map, hence it should return only single (sugared) version for every canonical type.

Reply:
Please see if you can clarify this comment a little. The mention of `resugar_map` here is also a little confusing, since it's implied that it's the only user of `TypeEnumerator`.

Also the mention of base class `TraverseType` below seems a little off now, because if you called `this->TraverseType(...)`, it would basically go into infinite recursion, right?

Let me know if you want help phrasing something.

Reply:
> Also the mention of base class TraverseType below seems a little off now, because if you called this->TraverseType(...), it would basically go into infinite recursion, right?

Yes, it would do, but avoiding of inserting multiple types with the same canonical type seems also important, as for me (please recall [the UB that I've mentioned](https://github.com/include-what-you-use/include-what-you-use/pull/1182#issuecomment-1384322467)).

I can think about some better wording for a while...

Reply:
I think the comment as stands now is good, but you have a much deeper understanding of this than I do. Does the UB still exist? If so, should we open a separate issue for it, so it's not forgotten?

Reply:
The described UB with more than one type from a single sugar chain inserted into `seen_types_` has been fixed here, I hope. But the issue with sibling types like in `Tpl<A, TypedefForA>` still remains. But I'm not sure that it will be solved in general in some day, because it is unknown in the common case to which template argument should be attributed some `A` usage inside the specialization.